crates/biome_cli/src/reporter/sarif.rs

```diff
@@ -0,0 +1,256 @@
+                        diagnostic_to_sarif(diagnostic)
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let report = SarifReport {
+            schema: "https://json.schemastore.org/sarif-2.1.0.json",
+            version: "2.1.0",
+            runs: Vec::from([SarifRun {
+                tool: SarifTool {
+                    driver: SarifDriver {
+                        name: "Biome",
+                        information_uri: "https://biomejs.dev",
+                        rules: sarif_rules,
+                    },
+                },
+                results: sarif_results,
+            }]),
+        };
+
+        let result = serde_json::to_string_pretty(&report)?;
+
+        self.0.log(markup! {
+            {result}
+        });
+
+        Ok(())
+    }
+}
+
+fn diagnostic_to_sarif<'a>(diagnostic: &'a Error) -> Option<SarifResult<'a>> {
+    let message = SarifResultMessage {
+        text: PrintDescription(diagnostic).to_string(),
+    };
+    let location = diagnostic.location();
+    let location = to_sarif_result_location(location);
+
+    Some(SarifResult {
+        rule_id: diagnostic
+            .category()
+            .map(|category| category.name())
+            .unwrap_or_default(),
+        level: match diagnostic.severity() {
+            Severity::Hint => "note",
+            Severity::Information => "note",
+            Severity::Warning => "warning",
+            Severity::Error => "error",
+            Severity::Fatal => "error",
+        },
+        message,
+        locations: Vec::from([location]),
+    })
+}
+
+fn category_to_sarif(category: &Category) -> SarifDriverRule {
+    let name = category.name();
+    let link = category.link().unwrap_or_default();
+
+    SarifDriverRule {
+        id: name,
+        short_description: SarifDriverRuleDescription { text: link },
+        full_description: SarifDriverRuleDescription { text: link },
+        help: SarifDriverRuleDescription { text: link },
+    }
+}
+
+fn to_sarif_result_location(location: Location) -> SarifResultLocation {
+    if let (Some(span), Some(source_code), Some(resource)) =
+        (location.span, location.source_code, location.resource)
+    {
+        let source = SourceFile::new(source_code);
+        let Some(start) = source.location(span.start()).ok() else {
+            return SarifResultLocation::default();
+        };
+        let Some(end) = source.location(span.end()).ok() else {
+            return SarifResultLocation::default();
+        };
+
+        SarifResultLocation {
+            physical_location: SarifResultLocationPhysicalLocation {
+                artifact_location: SarifResultLocationPhysicalLocationArtifactLocation {
+                    uri: if let Some(uri) = resource.as_file() {
+                        uri.to_string()
+                    } else {
+                        String::new()
+                    },
```

Comment:
Looking at [the example](https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md#a-simple-example), the file URL should be absolute. You should use the working directory and join the paths. I believe we do this in the gitlab reporter